### PR TITLE
JENKINS-61202 Make job configuration pages read-only when accessing them with the Extended Read permission

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-61202](https://issues.jenkins-ci.org/browse/JENKINS-61202).

Refer to discussion here:
https://github.com/jenkinsci/jenkins/pull/4149#issuecomment-582040203

<details>
<summary>Freestyle with configure</summary>

freestyle-configure
</details>

<details>
<summary>Freestyle with extended-read</summary>

freestyle-read-only
</details>

<details>
<summary>Pipeline with configure</summary>

pipeline-configure
</details>

<details>
<summary>Pipeline with extended-read</summary>

pipeline-read
</details>

Pipeline looks a bit funny, but that will need fixes in workflow-global-cps-plugin

Ref: https://github.com/jenkinsci/jenkins/pull/4149#issuecomment-582040203

### Proposed changelog entries

* Entry 1: JENKINS-61202 RFE: Users with extended read permission now get a more read-only looking UI.
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

File: core/src/main/resources/lib/form/hetero-list/hetero-list.js [hunk 1]
```diff
@@ -10,128 +10,139 @@ Behaviour.specify("DIV.hetero-list-container", 'hetero-list', -100, function(e)
...
-                    // and preferred ordering function, return the position in the array
-                    // the prospect should be inserted.
-                    // (for example 0 if it should be the first item)
-                    function findBestPosition(prospect,current,order) {
-                        function desirability(pos) {
-                            var count=0;
-                            for (var i=0; i<current.length; i++) {
-                                if ((i<pos) == (order(current[i])<=order(prospect)))
-                                    count++;
+        if (btn) {
```
Comment (timja) [inline core/src/main/resources/lib/form/hetero-list/hetero-list.js]: added a nil check so it doesn't fail if the button isn't present

File: core/src/main/resources/lib/form/hetero-list/hetero-list.js [hunk 2]
```diff
@@ -10,128 +10,139 @@ Behaviour.specify("DIV.hetero-list-container", 'hetero-list', -100, function(e)
...
+                    new YAHOO.util.Anim(nc, {
+                        opacity: {to: 1}
+                    }, 0.2, YAHOO.util.Easing.easeIn).animate();
 
-        if (e.hasClassName("one-each")) {
-            // does this container already has a configured instance of the specified descriptor ID?
-            function has(id) {
-                return Prototype.Selector.find(e.childElements(),"DIV.repeated-chunk[descriptorId=\""+id+"\"]")!=null;
-            }
+                    Behaviour.applySubtree(nc, true);
```
Comment (fqueiruga) [inline core/src/main/resources/lib/form/hetero-list/hetero-list.js]: Can you add a comment to this block to explain its intent?

File: core/src/main/resources/lib/form/number.jelly [hunk 3]
```diff
@@ -76,12 +76,25 @@ THE SOFTWARE.
...
-         name="${name}"
-         value="${value}"
-         type="number"
-         ATTRIBUTES="${attrs}" EXCEPT="field clazz" />
+
+  <j:choose>
+    <j:when test="${displayOnlyMode}">
+      <j:choose>
+        <j:when test="${empty(value)}">
+          <span class="jenkins-not-applicable">N/A</span>
```
Comment (fqueiruga) [inline core/src/main/resources/lib/form/number.jelly]: Would it make sense to extract this `<j:choose>` to its own jelly template?

File: core/src/main/resources/lib/form/radioBlock.jelly [hunk 4]
```diff
@@ -58,7 +58,7 @@ THE SOFTWARE.
 	  <td colspan="3">
       <label>
 	      <input type="radio" name="${name}" value="${value}"
-             class="radio-block-control block-control" checked="${checked?'true':null}" />
+             class="radio-block-control block-control" checked="${checked?'true':null}" disabled="${displayOnlyMode?'true':null}" />
```
Comment (fqueiruga) [inline core/src/main/resources/lib/form/radioBlock.jelly]: Any chance you could format this putting attributes in multiple lines, like on other inputs on this PR?

I'm also in favour of adding spacing within ternary operators. But maybe that's too much for this PR.

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 5]
```diff
@@ -1382,7 +1382,7 @@ function updateOptionalBlock(c,scroll) {
     while (!vg.hasClassName("rowvg-start"))
         vg = vg.next();
 
-    var checked = xor(c.checked,Element.hasClassName(c,"negative"));
+    var checked = xor(c.checked,Element.hasClassName(c,"negative")) || Element.hasClassName(c, "display-only-checked");
```
Comment (fqueiruga) [inline war/src/main/webapp/scripts/hudson-behavior.js]: Could `Element.hasClassName(c,"negative")) || Element.hasClassName(c, "display-only-checked"` be extracted into a variable? I think this conditional complicates the `xor` function a bit.

File: core/src/main/resources/lib/form/hetero-list/hetero-list.js (hunk 2 above)
Comment (timja) [reply core/src/main/resources/lib/form/hetero-list/hetero-list.js]: I have no idea it's intent, I added an if block to safe guard against an undefined value, the rest of the changes are all indentation, you can see the changes without white space related changes showing my using this url:
https://github.com/jenkinsci/jenkins/pull/4479/files?utf8=%E2%9C%93&diff=split&w=1

File: core/src/main/resources/lib/form/radioBlock.jelly (hunk 4 above)
Comment (timja) [reply core/src/main/resources/lib/form/radioBlock.jelly]: I went for minimal diff where possible, (although my IDE has fought me to a degree in some files)

happy to format this

File: core/src/main/resources/lib/form/hetero-list/hetero-list.js (hunk 1 above)
Comment (fqueiruga) [reply core/src/main/resources/lib/form/hetero-list/hetero-list.js]: It seems that everything's nested within this clause. I'd suggest to just return if the button is null to keep the function flat.

File: war/src/main/webapp/scripts/hudson-behavior.js (hunk 5 above)
Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js]: this isn't actually in use anymore, I need to revert it.

I was trying to not use the native controls and use icon's / images so that users don't get confused due to disabled controls not being that obvious.

But the javascript wasn't the easiest to change and I didn't want to break anything so I went back to using the native controls but forgot to revert this change

File: core/src/main/resources/lib/form/number.jelly (hunk 3 above)
Comment (timja) [reply core/src/main/resources/lib/form/number.jelly]: Will take a look at this later on, fixed the other comments up

**Comment by fqueiruga** [comment]
Gonna approve this because it looks good, and my comments are just proposed enhancements.

**Review by fqueiruga — APPROVED** [review]

**Comment by rsandell** [comment]
I feel you need to document `displayOnlyMode` somehow, there are plugins that provide their own `repeatable` for example and needs to be adapted to not add the remove/add buttons etc.

**Comment by rsandell** [comment]
But it's cool that it is based on a jelly var because plugins don't need to bump core version to add compatibility for it.

**Review by MRamonLeon — COMMENTED** [review]
Changes look good so far, but we need to test it locally to be able to approve. Hope you understand Timja

**Comment by timja** [comment]
@rsandell where would you expect the docs to be?

File: war/src/main/webapp/css/style.css [hunk 6]
```diff
@@ -1,18 +1,18 @@
 /*
  * The MIT License
- * 
+ *
```
Comment (daniel-beck) [inline war/src/main/webapp/css/style.css]: Most of the changes in this file are whitespace only.

File: core/src/main/resources/hudson/model/Job/configure.jelly [hunk 7]
```diff
@@ -30,6 +30,15 @@ THE SOFTWARE.
 
   <l:layout norefresh="true" type="one-column" permission="${it.EXTENDED_READ}" title="${%Config(it.displayName)}">
 
+    <j:choose>
+      <j:when test="${app.hasPermission(it.CONFIGURE)}">
+        <j:set var="displayOnlyMode" value="false" />
+      </j:when>
+      <j:otherwise>
+        <j:set var="displayOnlyMode" value="true" />
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/Job/configure.jelly]: This seems a weird choice of name given that both Extended Read and System Read use the term _Read_. I'm also more familiar with "read only" as a term.

Why was this term chosen?

File: core/src/main/resources/lib/form/password.jelly [hunk 8]
```diff
@@ -60,74 +60,88 @@ THE SOFTWARE.
...
+                      <g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
+                        <g transform="translate(-504.000000, -199.000000)" fill="#788594">
+                          <path
+                              d="M520.914667,209.666667 L511.466667,209.666667 L511.466667,206.619333 C511.466667,204.014 513.584667,201.895333 516.190667,201.895333 C518.796667,201.895333 520.914667,204.014 520.914667,206.619333 L520.914667,209.666667 Z M516.190667,223.381333 C514.514,223.381333 513.143333,222.01 513.143333,220.333333 C513.143333,218.657333 514.514,217.286 516.190667,217.286 C517.867333,217.286 519.238,218.657333 519.238,220.333333 C519.238,222.01 517.867333,223.381333 516.190667,223.381333 Z M516.190667,199 C511.984667,199 508.571333,202.414 508.571333,206.619333 L508.571333,209.666667 L507.048,209.666667 C505.372,209.666667 504,211.038 504,212.714667 L504,227.952667 C504,229.628667 505.372,231 507.048,231 L525.334,231 C527.01,231 528.380667,229.628667 528.380667,227.952667 L528.380667,212.714667 C528.380667,211.038 527.01,209.666667 525.334,209.666667 L523.81,209.666667 L523.81,206.619333 C523.81,202.414 520.396667,199 516.190667,199 Z"/>
```
Comment (daniel-beck) [inline core/src/main/resources/lib/form/password.jelly]: Really unnecessary whitespace change, cannot even be hidden through `?w=1` on GitHub.

File: core/src/main/resources/lib/form/password.jelly [hunk 9]
```diff
@@ -60,74 +60,88 @@ THE SOFTWARE.
...
+                         class="complex-password-field hidden-password-field setting-input ${attrs.checkUrl!=null?'validated ':''}${attrs.clazz}"
+                         name="${attrs.name ?: '_.'+attrs.field}"
+                         value="${value}"
+                         type="hidden"
+                         checkMethod="post"
+                         ATTRIBUTES="${attrs}" EXCEPT="field clazz value"/>
+                <div class="hidden-password-placeholder">
+                  <div class="hidden-password-legend">
+                    <svg width="20px" height="25px" viewBox="0 0 25 32" version="1.1"
+                         xmlns="http://www.w3.org/2000/svg">
```
Comment (daniel-beck) [inline core/src/main/resources/lib/form/password.jelly]: Really unnecessary whitespace change, cannot even be hidden through `?w=1` on GitHub.

**Review by daniel-beck — COMMENTED** [review]
Tested indirectly via your combined JEP-223/224 branch you sent me in Gitter.

Looks pretty cool, and I see the commonality between this PR, and the JEP-224 changes in the form elements, so there's a lot of overlap.

Thanks for taking care of this!

(You really need to configure your editor to ignore bad whitespace though 😉 )

File: core/src/main/resources/hudson/model/Job/configure.jelly (hunk 7 above)
Comment (timja) [reply core/src/main/resources/hudson/model/Job/configure.jelly]: Planning to rename it to read, naming is hard.

**Comment by timja** [comment]
Whitespace issues fixed

**Comment by timja** [comment]
Docs are here:
https://github.com/jenkins-infra/jenkins.io/pull/2859

File: core/src/main/resources/lib/form/number.jelly (hunk 3 above)
Comment (timja) [reply core/src/main/resources/lib/form/number.jelly]: Done, in all except password as it has a different default text rendering handling

**Review by Wadeck — COMMENTED** [review]
Code looks good, I want to try it manually ASAP. I have special concerns about the textarea "read-only mode" that seems to not take the new line into account and to specifically look at the behavior for the disabled fields, thinking about copy/pasting values (disable = no copy/paste, readonly = copy/paste possible).
Between disabled and readonly, it's important for sure on text fields (and you already cover them!), but I am not sure on the checkbox and other. Will check ;)

**Review by oleg-nenashev — APPROVED** [review]
Thanks @timja! I agree with @Wadeck 's comments above, but at the same time it os a net improvement being compared to the current UX ("you can edit anything, just cannot save it").  So I am fine with integrating it even if behavior in some components is sub-optimal.

One good thing in this pull request is that one does not need to bump the core dependency to pick up the `readOnlyMode` checks in their configurations and taglibs. Variable existence check should do the job.

**Comment by oleg-nenashev** [comment]
P.S: From what I see there is no Extended Read test coverage in https://github.com/jenkinsci/acceptance-test-harness . Maybe we should add some? Not a blocker for this PR

**Review by markjacksonfishing — APPROVED** [review]

**Comment by timja** [comment]
@Wadeck / @MRamonLeon would you be able to take a look at this please?

**Review by res0nance — APPROVED** [review]
LGTM

**Comment by daniel-beck** [comment]
On-holding this: As discussed with @timja out of band, preferred process is:

1. Merge https://github.com/jenkinsci/jenkins/pull/4506
2. Rebase/re-do this on top of that (only the `configure.jelly` change should be left)
3. Merge this

**Review by daniel-beck — CHANGES_REQUESTED** [review]
"ExtendedRead" is actually two separate permissions:

* Job/ExtendedRead
* Agent/ExtendedRead

Both are governed by the same system property.

The latter appears to be missing here. If deliberate, asking for an explanation, else for it to be added.

Thanks 😃

**Comment by timja** [comment]
Didn’t even know agent extended read existed, will take a look later on

**Comment by timja** [comment]
@daniel-beck the extended-read plugin doesn't enable the permission for agent...
https://github.com/jenkinsci/extended-read-permission-plugin/blob/master/src/main/java/hudson/plugins/extendedread/ExtendedReadPermissionEnabler.java#L17
(which is why I didn't know about it)


Do you know the background to this?
1. why is it not enabled?
2. why are both permissions still disabled by default? is it not time to just enable them?

**Comment by timja** [comment]
@daniel-beck does this permission (for agents) even work? I can't see the required code for it and I can't see any different to read permission

**Comment by timja** [comment]
It seems to be API only from what I can tell? ref: https://github.com/jenkinsci/jenkins/pull/855

**Review by daniel-beck — APPROVED** [review]
This is fine to merge as is on top of the System Read change.

We can follow up Agent/ExtendedRead UI later (no UI — no inconsistent style!), as well as the bug in extended-read-plugin (lolwut).

**Comment by daniel-beck** [comment]
The clock is now ticking: This is going in tomorrow(ish) unless someone complains.

**Comment by daniel-beck** [comment]
> why are both permissions still disabled by default? is it not time to just enable them?

Not everyone needs them, and adding permissions is generally a bit of a mess given the resulting UI brokenness with extra wide tables in matrix-auth and similar plugins. If Jesse sees this message, I'm sure he'll reference some decade old Jira issues explaining this in detail.

Same with WipeOutPermission, ArtifactsPermission permission (the latter of which is very useful), except they're even weirder: Not implied by any permission 😆

**Comment by daniel-beck** [comment]
Potentially an `lts-candidate` given that all the controls from System Read are in and this ended up being a pretty small change.

**Comment by oleg-nenashev** [comment]
Somebody should create a Jira issue then

**Comment by timja** [comment]
> > why are both permissions still disabled by default? is it not time to just enable them?
> 
> Not everyone needs them, and adding permissions is generally a bit of a mess given the resulting UI brokenness with extra wide tables in matrix-auth and similar plugins. If Jesse sees this message, I'm sure he'll reference some decade old Jira issues explaining this in detail.
> 
> Same with WipeOutPermission, ArtifactsPermission permission (the latter of which is very useful), except they're even weirder: Not implied by any permission 😆

http://jenkins-ci.361315.n4.nabble.com/Allowing-read-only-access-to-config-xml-tp393245p393249.html

> KK
> I like your approach.  I feel like the system property is too esoteric> though, since the read-only view of configurations is pretty widely> desired.  Could it be moved to a config option on the global config page?> This would allow already started instances to enable this easily.Adding a switch to toggle the appearance of one more permission kinda defeats the purpose of reducing the configuration, I think.The thing I like about a system property is that people who really need the feature can start using it right away, and we get some time to decide whether it should be available to everyone by default or not.

It's been 10.5 years :joy:

**Comment by daniel-beck** [comment]
> decade old

Called it.

File: core/src/main/resources/hudson/model/Job/configure.jelly [hunk 10]
```diff
@@ -30,6 +30,15 @@ THE SOFTWARE.
 
   <l:layout norefresh="true" type="one-column" permission="${it.EXTENDED_READ}" title="${%Config(it.displayName)}">
 
+    <j:choose>
+      <j:when test="${app.hasPermission(it.CONFIGURE)}">
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/Job/configure.jelly]: https://issues.jenkins-ci.org/browse/JENKINS-61321 / https://github.com/jenkinsci/jenkins/pull/4541

Comment (timja) [reply core/src/main/resources/hudson/model/Job/configure.jelly]: 🤕, tested it with an admin user and read only, but not a lower privileged but still configure -.-
